### PR TITLE
NIFI-8748 Corrected PutKudu String to java.sql.Date parsing

### PR DESCRIPTION
**Pull request by exceptionfactory**
#### Description of PR

NIFI-8748 Corrects `PutKudu` Date field parsing when converting from a `String` Record field value to a `java.sql.Date` for a Kudu `PartialRow`.

The `DataTypeUtils.getDateFormat()` method sets a default time zone of `GMT` on the `SimpleDateFormat` used for parsing, which results in subtracting one day from the input date string.  Changing the approach to use a `SimpleDateFormat` with the default system time zone avoids the implicit date conversion and preserves the original date value following the conversion to `java.sql.Date`.  The time zone setting is not necessary since the date field does not include precision beyond a day boundary.

Changes include a new unit test method to verify the corrected behavior.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [X] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [X] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [X] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [X] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [X] Have you written or updated unit tests to verify your changes?
- [X] Have you verified that the full build is successful on JDK 8?
- [X] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by pvillard31** [comment]
@granthenke - can you eyeball this PR? it looks good to me but would appreciate your feedback :)

**Review by kevdoran — APPROVED** [review]
@exceptionfactory, thanks for this fix! LGTM based on the description in the PR/Jira. +1

File: nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/AbstractKuduProcessor.java [hunk 1]
```diff
@@ -412,6 +415,26 @@ protected void buildPartialRow(Schema schema, PartialRow row, Record record, Lis
...
+        return DataTypeUtils.toDate(value, () -> getDateFormat(), recordFieldName);
+    }
+
+    /**
+     * Get Date Format using Date Record Field default pattern and system time zone to avoid unnecessary conversion
+     *
+     * @return Date Format used to parsing date fields
+     */
+    private DateFormat getDateFormat() {
+        return new SimpleDateFormat(RecordFieldType.DATE.getDefaultFormat());
```
Comment (granthenke) [inline nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/AbstractKuduProcessor.java:435]: Should this use `fieldType.getFormat()` similar to the handling for Timestamp/UNIXTIME_MICROS above?

**Comment by granthenke** [comment]
Looks good to me, clearly an important fix. I just had a question about the use of the default format vs fieldType.getFormat().

File: nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/AbstractKuduProcessor.java (hunk 1 above)
Comment (exceptionfactory) [reply nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/AbstractKuduProcessor.java]: Thanks for the suggestion @granthenke, that seems like a helpful improvement as opposed to hard-coded default format of `yyyy-MM-dd` from `RecordFieldType.DATE.getDefaultFormat()`.  I updated the methods to accept the format from `DataType.getFormat()`.

**Comment by granthenke** [comment]
👍

**Comment by pvillard31** [comment]
Merged, thanks all!

**Pull request closed without merging**